### PR TITLE
ERC20BridgeSampler

### PR DESCRIPTION
**Pull request by dorothy-zbornak**
mukbang

## Description

Here's the sampler contract for the ERC20Bridge aggregator!

This facilitates the MVP of the aggregator, which just agnostically samples DEX quotes for off-chain sorting, and does not handle any funds. A lot of functions do just wrap each other in order to reduce the number of RPC calls required in getting a quote. Also, Kyber is not supported for buys right now.

These contracts live in their own package under `/contracts`, but I kind of think `extensions` might be a better fit, since it's such a one-off. Unfortunately, `extensions` is broken rn so maybe we can move it in there when they're migrated :woman_shrugging:.

## Checklist:

-   [ ] Prefix PR title with `[WIP]` if necessary.
-   [ ] Add tests to cover changes as needed.
-   [ ] Update documentation as needed.
-   [ ] Add new entries to the relevant CHANGELOG.jsons.

### CODE REVIEW

File: contracts/erc20/contracts/src/LibERC20Token.sol [hunk 1]
```diff
@@ -91,6 +92,21 @@ library LibERC20Token {
         _callWithOptionalBooleanResult(token, callData);
     }
 
+    /// @dev Retrieves the number of decimals for a token.
+    ///      Returns `18` if the call reverts.
+    /// @return The number of decimals places for the token.
+    function decimals(address token)
```
Comment (dorothy-zbornak) [inline contracts/erc20/contracts/src/LibERC20Token.sol]: There seems to be some confusion about whether `decimals` is part of the standard (I think it is), so I added this just to be safe.

File: contracts/erc20-bridge-sampler/test/erc20-bridge-sampler.ts [hunk 2]
```diff
@@ -0,0 +1,821 @@
...
+        });
+
+        it('throws with non-ERC20 maker asset data', async () => {
+            const tx = testContract
+                .queryOrdersAndSampleSells(
+                    createOrders(MAKER_TOKEN, TAKER_TOKEN).map(o => ({
+                        ...o,
+                        makerAssetData: INVALID_ASSET_PROXY_ASSET_DATA,
+                    })),
+                    [...SELL_SOURCES.map(n => allSources[n]), randomAddress()],
```
Comment (jalextowle) [inline contracts/erc20-bridge-sampler/test/erc20-bridge-sampler.ts]: Nit:
```suggestion
                    SELL_SOURCES.map(n => allSources[n]),
```

This should also be changed in a few places below.

File: contracts/erc20-bridge-sampler/test/erc20-bridge-sampler.ts [hunk 3]
```diff
@@ -0,0 +1,821 @@
...
+        });
+
+        it('throws with non-ERC20 maker asset data', async () => {
+            const tx = testContract
+                .queryOrdersAndSampleBuys(
+                    createOrders(MAKER_TOKEN, TAKER_TOKEN).map(o => ({
+                        ...o,
+                        makerAssetData: INVALID_ASSET_PROXY_ASSET_DATA,
+                    })),
+                    [...BUY_SOURCES.map(n => allSources[n]), randomAddress()],
```
Comment (jalextowle) [inline contracts/erc20-bridge-sampler/test/erc20-bridge-sampler.ts]: Same as above.

File: contracts/erc20-bridge-sampler/.npmignore [hunk 4]
```diff
@@ -0,0 +1,10 @@
+# Blacklist all files
+.*
+*
+# Whitelist lib
+!lib/**/*
```
[the remark this answers is not in the document]
Comment (dorothy-zbornak) [reply contracts/erc20-bridge-sampler/.npmignore]: This was copy & pasted from `integrations` :woman_shrugging:

File: contracts/erc20-bridge-sampler/test/erc20-bridge-sampler.ts [hunk 5]
```diff
@@ -0,0 +1,851 @@
...
+            .plus(_minRate)
+            .div(RATE_DENOMINATOR);
+    }
+
+    function getDeterministicTokenDecimals(token: string): number {
+        if (token === WETH_ADDRESS) {
+            return 18;
+        }
+        return (
+            (new BigNumber(getPackedHash(token)).mod(MAX_DECIMALS - MIN_DECIMALS).toNumber() as number) + MIN_DECIMALS // linter is confused
```
Comment (mzhu25) [inline contracts/erc20-bridge-sampler/test/erc20-bridge-sampler.ts]: wot

Comment (dorothy-zbornak) [reply contracts/erc20-bridge-sampler/test/erc20-bridge-sampler.ts]: For some reason `toNumber()` wasn't being picked up as a `number`.

Comment (dorothy-zbornak) [reply contracts/erc20-bridge-sampler/test/erc20-bridge-sampler.ts]: I'll clean it up in the rebase.

**Pull request merged**